service/internal/ebs/ebs.go

```diff
@@ -208,25 +208,60 @@ func (e *EBS) ListVolumes(cr infrastructurev1alpha2.AWSCluster, filterFuncs ...f
 		if !IsFiltered(v, filterFuncs) {
 			continue
 		}
-
+		ignoreVolume := false
 		attachments := []VolumeAttachment{}
 
 		if len(v.Attachments) > 0 {
 			for _, a := range v.Attachments {
+
+				if *a.InstanceId != "" {
+					i := &ec2.DescribeInstancesInput{
+						InstanceIds: aws.StringSlice([]string{*a.InstanceId}),
+					}
+
+					o, err := e.client.DescribeInstances(i)
+					if err != nil {
+						return nil, microerror.Mask(err)
+					}
+
+					// if the instance does not have the proper cluster-id tag than ignore the volume
+					if len(o.Reservations) > 0 && len(o.Reservations[0].Instances) > 0 &&
+						!instanceHasTag(o.Reservations[0].Instances[0], key.TagCluster, cr.Labels[key.TagCluster]) {
+						e.logger.LogCtx(ctx, "level", "warning", "message", fmt.Sprintf("EBS volume %#q is attached to instance %#q which does not belong to the cluster %#q, ignoring the volume", *a.VolumeId, *a.InstanceId, cr.Labels[key.TagCluster]))
+
+						// do not add this volume to the volume list
+						// this volume will be ignored by the operator
+						ignoreVolume = true
+						break
+					}
+				}
+
 				attachments = append(attachments, VolumeAttachment{
 					Device:     *a.Device,
 					InstanceID: *a.InstanceId,
 				})
 			}
 		}
 
-		volume := Volume{
-			VolumeID:    *v.VolumeId,
-			Attachments: attachments,
-		}
+		if !ignoreVolume {
+			volume := Volume{
+				VolumeID:    *v.VolumeId,
+				Attachments: attachments,
+			}
 
-		volumes = append(volumes, volume)
+			volumes = append(volumes, volume)
+		}
```

Comment:
I wonder whats better to read that or this
```
if ignoreVolume {
  continue
}
```

Reply:
An early continue makes sense to me.